crates/pathfinder/src/cairo/ext_py/sub_process.rs

```diff
@@ -217,10 +225,79 @@ async fn spawn(
                     }
                 }
 
-                trace!("{:?}", buffer.trim());
+                let buffer = buffer.trim();
+
+                struct InvalidLevel;
+
+                // first one is the level selector, assuming this is code controlled by us
+                let mut chars = buffer.chars();
+                let level = match chars.next() {
+                    Some('0') => Ok(tracing::Level::ERROR),
+                    Some('1') => Ok(tracing::Level::WARN),
+                    Some('2') => Ok(tracing::Level::INFO),
+                    Some('3') => Ok(tracing::Level::DEBUG),
+                    Some('4') => Ok(tracing::Level::TRACE),
+                    Some(_) => Err(InvalidLevel),
+                    None => {
+                        // whitespace only line
+                        continue;
+                    }
+                };
+
+                let (level, displayed) = if let Ok(level) = level {
+                    let rem = chars.as_str();
+                    // if we treat the thing as json, we can easily get multiline messages, which most
+                    // are
+                    let displayed = if rem.starts_with('"') {
+                        serde_json::from_str::<String>(rem)
+                            .ok()
+                            .map(std::borrow::Cow::Owned)
+                    } else {
+                        None
+                    };
+
+                    (level, displayed.unwrap_or(std::borrow::Cow::Borrowed(rem)))
+                } else {
+                    // this means that the line probably comes from beyond our code
+                    (tracing::Level::ERROR, std::borrow::Cow::Borrowed(buffer))
+                };
+
+                // if the python script would turn out to be very chatty, this would become an issue
+                let current_span = current_span.lock().unwrap_or_else(|e| e.into_inner());
+
+                // not sure if this makes sense, but use the outer span if there's no specific
+                // span at the moment, otherwise use the specific one.
+                //
+                // the hope is that for each stderr message there will be either hierarchy:
+                //
+                // +         span from ext_py::service
+                // |
+                // ---       span at `async fn spawn` time (has pid)
+                //   |
+                //   +----  `span` here, which has just stderr
+                //
+                // +         user code calling ext_py::Handle
+                // |           - pid gets populated at ext_py::sub_process
+                // +-+       `*g` here, which has just stderr
+                //
+                let used_span = if current_span.is_none() {
+                    &default_span
+                } else {
+                    &*current_span
+                };
```

Comment:
This I think is the most important part, that this comment is understandable. I think the `default_span` and `current_span` are quite good but I didn't think they were sufficient. I did however write the comment before ending with these names.

Reply:
No direct calls, this is a spawned task which will just read the stderr until it can't. `ext_py::Handle` will push work to the queue which sub_process tasks read through `Arc<Mutex<Receiver<_>>>`. 

The span from the original call to the `Handle::{call,estimate_fee}` is propagated to the subprocess first by the work queue, then further to the subprocesses' stderr reading task through this `current_span` which is another `Arc<Mutex<Span>>`.

I am thinking this comment is extra.

> Or is that for the examples?

Not sure what do you mean here. Examples of getting the span dragged here are best seen when [this tx span](https://github.com/eqlabs/pathfinder/blob/484e459b217128dc1452eeaf64a479b52b1e1c01/crates/pathfinder/examples/estimate_past_transactions.rs#L192) is propagated down to the task forwarding stderr, as seen in this output:

![image](https://user-images.githubusercontent.com/1081634/176429598-390dcf1f-612e-47eb-a1f4-805d600fe71f.png)